File: 130-surrounded-regions/surrounded-regions.cpp
class Solution {
public:
    void dfs(int r, int c, vector<vector<char>> &board, vector<vector<bool>> &vis){
        int n = board.size();
        int m = board[0].size();
        vis[r][c] = true;
        int dx[] ={-1,0,1,0};
        int dy[] ={0,-1,0,1};

        for(int d = 0; d < 4; d++){
            int rr = r + dx[d];
            int cc = c + dy[d];

            if(rr>=0 && cc>=0 && rr < n && cc < m && !vis[rr][cc] && board[rr][cc] == 'O' ){
                dfs(rr, cc, board, vis);
            }
        }
    }
    void solve(vector<vector<char>>& board) {
        //BY DFS and runnning through all borders and finding their connected Os and marking them
        //except them all the Os will be converted to X
        //TC - O(n + n + 4*n*m) and SC - O(n * m)
        int n = board.size();
        int m = board[0].size();
        vector<vector<bool>> vis(n, vector<bool>(m,false));
        for(int j = 0; j < m; j++){
            if(!vis[0][j] && board[0][j] == 'O'){
                dfs(0, j, board, vis);
            }
            if(!vis[n-1][j] && board[n-1][j] == 'O'){
                dfs(n-1, j, board, vis);
            }
        }

        for(int i = 0; i< n; i++){
            if(!vis[i][0] && board[i][0] == 'O'){
                dfs(i, 0, board, vis);
            }
            if(!vis[i][m-1] && board[i][m-1] == 'O'){
                dfs(i, m-1, board, vis);
            }
        }

        for(int i = 0; i <n; i++){
            for(int j = 0; j < m; j++){
                if(!vis[i][j]){
                    board[i][j] = 'X';
                }
            }
        }
    }
};

File: 130-surrounded-regions/surrounded-regions.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        n = len(board)
        m = len(board[0])

        vis = [[0]*m for _ in range(n)]        
        q = deque()

        for i in range(n):
            if board[i][0] == 'O':
                vis[i][0] = 1
                q.append((i,0))
            if board[i][m-1] == 'O':
                vis[i][m-1] = 1
                q.append((i,m-1))

        for j in range(m):
            if board[0][j] == 'O':
                vis[0][j] = 1
                q.append((0,j))
            if board[n-1][j] == 'O':
                vis[n-1][j] = 1
                q.append((n-1,j))
        dx = [0,0,1,-1]
        dy = [1,-1,0,0]
        while q:
            r,c = q.popleft()
            for d in range(4):
                rr = r + dx[d]
                cc = c + dy[d]
                if rr>=0 and rr < n and cc>0 and cc < m and vis[rr][cc]==0 and board[rr][cc] == 'O':
                    vis[rr][cc] = 1
                    q.append((rr,cc))

        for i in range(n):
            for j in range(m):
                if board[i][j] == 'O' and vis[i][j]==0:
                    board[i][j] = 'X'
    